File: CPP/ByRound/1823/1823_E_Removing_Graph.cpp
// Problem: CF 1823 E - Removing Graph
// https://codeforces.com/contest/1823/problem/E

/*
 * Problem: E. Removing Graph
 * 
 * Purpose: Determine the winner of a game played on a graph where each vertex has degree 2.
 *          Players take turns removing connected subgraphs of size between l and r.
 *          The game is about the nim-sum of cycle lengths in the graph.
 *
 * Algorithm:
 * - Each connected component in the graph is a cycle since every vertex has degree 2.
 * - For each cycle of size s, we calculate its contribution to the nim-sum:
 *   - If s >= l + r, the cycle contributes 0 to the nim-sum.
 *   - Otherwise, the cycle contributes (s / l) to the nim-sum.
 * - If the total nim-sum is non-zero, Alice wins; otherwise, Bob wins.
 *
 * Time Complexity: O(n * α(n)) where α is the inverse Ackermann function (n forDSU operations)
 * Space Complexity: O(n) for storing DSU structure
 */

#include <stdio.h>
#include <algorithm>
#include <iostream>
#include <utility>
#include <vector>

using namespace std;
struct Dsu {
  vector<int> dsu_vt;
  vector<int> sz_vt;
  Dsu() { init(0); }
  Dsu(int size) { init(size); }
  void init(int size) {
    dsu_vt.resize(size + 1);
    sz_vt.resize(size + 1);
    for (int i = 0; i <= size; i++)
      dsu_vt[i] = i, sz_vt[i] = 1;
  }
  int find(int u) { return dsu_vt[u] ^ u ? dsu_vt[u] = find(dsu_vt[u]) : u; }
  int size(int u) { return sz_vt[find(u)]; }
  void merge(int u, int v) {
    int ru = find(u), rv = find(v);
    if (ru != rv) {
      if (ru < rv)
        swap(ru, rv);
      dsu_vt[ru] = rv;
      sz_vt[rv] += sz_vt[ru];
    }
  }
};
void solve(int n, int l, int r) {
  Dsu dsu(n);
  for (int i = 1, u, v; i <= n; i++) {
    scanf("%d%d", &u, &v);
    dsu.merge(u, v);
  }
  int ans = 0;
  for (int i = 1; i <= n; i++)
    if (dsu.find(i) == i)
      ans ^= (dsu.size(i) >= l + r ? 0 : dsu.size(i) / l); // Compute nim-sum
  printf("%s\n", ans > 0 ? "Alice" : "Bob");
}
int main() {
  int n, l, r;
  while (cin >> n >> l >> r)
    solve(n, l, r);
  return 0;
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByTier/E/1823_E_Removing_Graph.py
# Problem: CF 1823 E - Removing Graph
# https://codeforces.com/contest/1823/problem/E

"""
Algorithm: Game Theory with DSU (Disjoint Set Union)
This problem involves a game on a graph where each vertex has degree 2, meaning the graph consists of cycles and paths.
The key insight is that we can model this using DSU to find connected components, then calculate the nimber (XOR) of each component's size.
For each component of size x, if x < L + R, we compute x // L and XOR it into the final result.
If the final XOR is 0, Bob wins; otherwise, Alice wins.
Time Complexity: O(N * α(N)) where α is the inverse Ackermann function, due to DSU operations
Space Complexity: O(N) for DSU and auxiliary data structures
"""
import os
import sys
from collections import defaultdict as ddic
from io import BytesIO, IOBase


class DSU:
    def __init__(self, N):
        self.par = list(range(N))  # Parent array for DSU
        self.sz = [1] * N          # Size of each component

    def find(self, x):
        # Path compression for efficiency
        if self.par[x] != x:
            self.par[x] = self.find(self.par[x])
        return self.par[x]

    def union(self, x, y):
        # Union by size for efficiency
        xr, yr = self.find(x), self.find(y)
        if xr == yr:
            return False
        if self.sz[xr] < self.sz[yr]:
            xr, yr = yr, xr
        self.par[yr] = xr
        self.sz[xr] += self.sz[yr]
        self.sz[yr] = self.sz[xr]
        return True

    def size(self, x):
        return self.sz[self.find(x)]


def solve():
    N, L, R = readlist()  # Read n, l, r
    dsu = DSU(N)
    # Read all edges and build the DSU structure
    for _ in range(N):
        u, v = readlist()
        u -= 1  # Convert to 0-indexed
        v -= 1
        dsu.union(u, v)
    
    # Group vertices by their component root and count sizes
    sizes = ddic(int)
    for u in range(N):
        sizes[dsu.find(u)] += 1
    
    sizes = list(sizes.values())  # Extract component sizes
    
    ans = 0
    # For each component, calculate contribution to nimber
    for x in sizes:
        if x < L + R:
            ans ^= x // L  # XOR the quotient into final answer
    return ans


def main():
    T = 1

    for tc in range(T):
        ans = solve()

        print("Alice" if ans else "Bob")  # Alice wins if xor is non-zero


BUFSIZE = 8192


class FastIO(IOBase):
    newlines = 0

    def __init__(self, file):
        self._file = file
        self._fd = file.fileno()
        self.buffer = BytesIO()
        self.writable = "x" in file.mode or "r" not in file.mode
        self.write = self.buffer.write if self.writable else None

    def read(self):
        while True:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            if not b:
                break
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines = 0
        return self.buffer.read()

    def readline(self):
        while self.newlines == 0:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            self.newlines = b.count(b"\n") + (not b)
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines -= 1
        return self.buffer.readline()

    def flush(self):
        if self.writable:
            os.write(self._fd, self.buffer.getvalue())
            self.buffer.truncate(0), self.buffer.seek(0)


class IOWrapper(IOBase):
    def __init__(self, file):
        self.buffer = FastIO(file)
        self.flush = self.buffer.flush
        self.writable = self.buffer.writable
        self.write = lambda s: self.buffer.write(s.encode("ascii"))
        self.read = lambda: self.buffer.read().decode("ascii")
        self.readline = lambda: self.buffer.readline().decode("ascii")


sys.stdin, sys.stdout = IOWrapper(sys.stdin), IOWrapper(sys.stdout)
input = lambda: sys.stdin.readline().rstrip("\r\n")
readint = lambda: int(input())
readlist = lambda: list(map(int, input().split()))


if __name__ == "__main__":
    main()


# https://github.com/VaHiX/CodeForces/